Samples/AudioCreation/cs/AudioCreation/Scenario7_RecordPlayback.xaml.cs

```diff
@@ -0,0 +1,570 @@
+                }
+
+                fixed (byte* buf = byteBuffer)
+                {
+                    for (int i = 0; i < capacityInBytes; i++)
+                    {
+                        byte b = dataInBytes[i + bufferStart];
+                        buf[i + maxIndex] = b;
+
+                        if (b != 0)
+                        {
+                            nonZeroSampleCount++;
+                        }
+                    }
+                }
+                maxIndex += capacityInBytes;
+            }
+
+            audioFrameCount++;
+        }
+
+        private void UpdateStatusWhileRecording()
+        {
+            uint max = maxIndex >> 3;
+            uint curr = currentIndex >> 3;
+            int frameCount = audioFrameCount;
+            // may have stopped recording while this async notification was in flight
+            if (isRecording)
+            {
+                rootPage.NotifyUser($"recording: max samples {max}, frame count {frameCount}, last cap {lastCapacityEncountered}, max cap {maxCapacityEncountered}", NotifyType.StatusMessage);
+            }
+        }
+
+        private void FrameInputNode_QuantumStarted(AudioFrameInputNode sender, FrameInputNodeQuantumStartedEventArgs args)
+        {
+            uint requiredSamples = (uint)args.RequiredSamples;
+
+            if (requestStartPlaying)
+            {
+                Debug.Assert(!isRecording);
+                Debug.Assert(!requestStartRecording);
+                Debug.Assert(!isPlaying);
+
+                isPlaying = true;
+                requestStartPlaying = false;
+                currentIndex = 0;
+            }
+
+            if (isPlaying)
+            {
+                Debug.Assert(!isRecording);
+                Debug.Assert(!requestStartPlaying);
+                Debug.Assert(!requestStartRecording);
+
+                HandleOutgoingAudio(requiredSamples);
+
+                /*
+                int currentStatusMsec = DateTime.Now.Millisecond;
+                // update status every tenth of a second
+                if (lastStatusMsec % 100 > currentStatusMsec % 100)
+                {
+                    CoreApplication.MainView.CoreWindow.Dispatcher.TryRunAsync(
+                        Windows.UI.Core.CoreDispatcherPriority.Normal,
+                        UpdateStatusWhilePlaying);
+                }
+                lastStatusMsec = currentStatusMsec;
+                */
+            }
+        }
+
+        // We keep a 1MB AudioFrame and reuse it.
+        static AudioFrame s_audioFrame = new Windows.Media.AudioFrame(1024 * 1024);
+        private int s_zeroByteOutgoingFrameCount;
+
+        /// <summary>
+        /// Handle a frame of outgoing audio to the output device.
+        /// </summary>
+        /// <remarks>
+        /// This pushes a megabyte of audio each time, to greatly reduce the number of callbacks.
+        /// 
+        /// We don't need the requiredSamples argument but we don't 
```

Comment:
And actually, I tried experimenting with using requiredSamples, and realized I'm not sure what code pattern to use.  Currently I reuse the single one-megabyte AudioFrame.  The FrameInputNode API seems to have no way to add just a *portion* of an AudioFrame.  In other words, if I get passed a requiredSamples value of 480 (3,840 bytes), am I supposed to create a new AudioFrame exactly 3,840 bytes in size?  Creating objects on every callback in C# is a good way to make too much garbage, so I was hoping to avoid it.

Is there something I am overlooking?  What code pattern would you suggest here?  Thanks.